### PR TITLE
Supervised contrastive learning

### PR DESCRIPTION
**Pull request by divyashreepathihalli**
Verified on colab - https://colab.sandbox.google.com/drive/1GlJzZiFQuhdBO-iNeB7Km9JXZy_W6YPG#scrollTo=EfMfko5kd8pP

### CODE REVIEW

**Comment by fchollet** [comment]
Thanks for the PR! Can you post the diff?

**Comment by divyashreepathihalli** [comment]
> Thanks for the PR! Can you post the diff?

here is the diff - https://github.com/keras-team/keras/pull/18591/commits/dbdea7bffc77f0bd4d4088bc3cda8a17151e8c03

I will update the number of epochs to 50.

File: examples/keras_io/vision/supervised_contrastive_learning.py [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+As shown in the experiments, using the supervised contrastive learning technique
+outperformed the conventional technique in terms of the test accuracy. Note that
+the same training budget (i.e., number of epochs) was given to each technique.
+Supervised contrastive learning pays off when the encoder involves a complex
+architecture, like ResNet, and multi-class problems with many labels.
+In addition, large batch sizes and multi-layer projection heads
+improve its effectiveness. See the [Supervised Contrastive Learning](https://arxiv.org/abs/2004.11362)
+paper for more details.
+
+You can use the trained model hosted on [Hugging Face Hub](https://huggingface.co/keras-io/supervised-contrastive-learning-cifar10)
```
Comment (fchollet) [inline examples/keras_io/vision/supervised_contrastive_learning.py:238]: You can remove the HuggingFace references.

File: keras/losses/losses.py [hunk 2]
```diff
@@ -1858,3 +1858,41 @@ def binary_focal_crossentropy(
...
     return ops.mean(focal_bce, axis=axis)
+
+
+@keras_export(
+    [
+        "keras.metrics.npairs_loss",
+        "keras.losses.npairs_loss",
+    ]
+)
+def npairs_loss(y_true, y_pred):
```
Comment (fchollet) [inline keras/losses/losses.py:1869]: Let's not add it to the API -- keep the implementation in the example code.

File: examples/keras_io/vision/supervised_contrastive_learning.py [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+
+
+class SupervisedContrastiveLoss(keras.losses.Loss):
+    def __init__(self, temperature=1, name=None):
+        super().__init__(name=name)
+        self.temperature = temperature
+
+    def __call__(self, labels, feature_vectors, sample_weight=None):
+        # Normalize feature vectors
+        feature_vectors_normalized = tf.math.l2_normalize(feature_vectors, axis=1)
```
[the remark this answers is not in the document]
Comment (divyashreepathihalli) [reply examples/keras_io/vision/supervised_contrastive_learning.py]: done!

File: keras/losses/losses.py (hunk 2 above)
Comment (divyashreepathihalli) [reply keras/losses/losses.py]: done

**Comment by divyashreepathihalli** [comment]
@fchollet here is the diff for original code to converted code - https://github.com/keras-team/keras/pull/18591/commits/1e793a3de49ed686af5d680f9640273bd029c075

File: examples/keras_io/vision/supervised_contrastive_learning.py [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+"""
+"""
+## Setup
+"""
+
+import keras
+from keras import ops
+from keras import layers
+from keras.applications.resnet_v2 import ResNet50V2
+from keras.losses.loss import squeeze_to_same_rank
```
Comment (fchollet) [inline examples/keras_io/vision/supervised_contrastive_learning.py:32]: This is a private API, we can't use it here.

File: examples/keras_io/vision/supervised_contrastive_learning.py [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+      y_pred: Predicted values of shape `[batch_size, batch_size]` of
+        similarity matrix between embedding matrices.
+
+    Returns:
+      npairs_loss: float scalar.
+    """
+    y_pred = ops.cast(y_pred, "float32")
+    y_true = ops.cast(y_true, y_pred.dtype)
+    y_true = ops.expand_dims(y_true, -1)
+    y_true, y_pred = squeeze_to_same_rank(y_true, y_pred)
```
Comment (fchollet) [inline examples/keras_io/vision/supervised_contrastive_learning.py:152]: You probably don't need this statement (or the one above)

File: examples/keras_io/vision/supervised_contrastive_learning.py [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+## Setup
+"""
+
+import keras
+from keras import ops
+from keras import layers
+from keras.applications.resnet_v2 import ResNet50V2
+from keras.losses.loss import squeeze_to_same_rank
+import os
+os.environ["KERAS_BACKEND"] = "tensorflow"
```
Comment (fchollet) [inline examples/keras_io/vision/supervised_contrastive_learning.py:34]: If it works with all backends you can remove this.

Comment (divyashreepathihalli) [reply examples/keras_io/vision/supervised_contrastive_learning.py]: done

**Review by fchollet — COMMENTED** [review]
Thanks for the update! The code looks good! Can you confirm that it runs as expected with TF, JAX, and torch?

**Comment by divyashreepathihalli** [comment]
> Thanks for the update! The code looks good! Can you confirm that it runs as expected with TF, JAX, and torch?

Here is a colab - https://colab.sandbox.google.com/drive/1GlJzZiFQuhdBO-iNeB7Km9JXZy_W6YPG#scrollTo=EfMfko5kd8pP
I verified that it works on all 3 backends.

**Review by fchollet — APPROVED** [review]
LGTM, thank you

**Pull request merged**